File: patterns/Creational/Builder/JavaScript/RealWorldExample/ComputerAssembly.js
// Builder Design Pattern - Computer Assembly
//
// This example demonstrates the Builder Design Pattern for assembling
// different types of computers with varying configurations (Gaming, Office).

// ComputerBuilder interface declares the product building steps that are common
// for all types of builders.
class ComputerBuilder {
    setCPU(cpu) {}
    setRAM(ram) {}
    setStorage(storage) {}
    setGraphicsCard(gpu) {}
    getComputer() {}
}

// Concrete Builder for assembling a Gaming Computer.
class GamingComputerBuilder extends ComputerBuilder {
    constructor() {
        super();
        this.computer = new Computer();
    }

    setCPU(cpu) {
        this.computer.parts['CPU'] = cpu;
        return this;
    }

    setRAM(ram) {
        this.computer.parts['RAM'] = ram;
        return this;
    }

    setStorage(storage) {
        this.computer.parts['Storage'] = storage;
        return this;
    }

    setGraphicsCard(gpu) {
        this.computer.parts['GraphicsCard'] = gpu;
        return this;
    }

    getComputer() {
        return this.computer;
    }
}

// Concrete Builder for assembling an Office Computer.
class OfficeComputerBuilder extends ComputerBuilder {
    constructor() {
        super();
        this.computer = new Computer();
    }

    setCPU(cpu) {
        this.computer.parts['CPU'] = cpu;
        return this;
    }

    setRAM(ram) {
        this.computer.parts['RAM'] = ram;
        return this;
    }

    setStorage(storage) {
        this.computer.parts['Storage'] = storage;
        return this;
    }

    setGraphicsCard(gpu) {
        this.computer.parts['GraphicsCard'] = 'Integrated GPU'; // Office computers don't need high-end GPUs
        return this;
    }

    getComputer() {
        return this.computer;
    }
}

// Product class representing a Computer.
class Computer {
    constructor() {
        this.parts = {};
    }

    showParts() {
        console.log('Computer Configuration:');
        for (const [key, value] of Object.entries(this.parts)) {
            console.log(`${key}: ${value}`);
        }
    }
}

// Director class guides the building process.
class Director {
    buildGamingComputer(builder) {
        return builder.setCPU('Intel i9')
            .setRAM('32GB')
            .setStorage('1TB SSD')
            .setGraphicsCard('NVIDIA RTX 3090')
            .getComputer();
    }

    buildOfficeComputer(builder) {
        return builder.setCPU('Intel i5')
            .setRAM('16GB')
            .setStorage('500GB SSD')
            .setGraphicsCard('Integrated GPU')
            .getComputer();
    }
}

// Client code.
function clientCode(director) {
    const gamingBuilder = new GamingComputerBuilder();
    const officeBuilder = new OfficeComputerBuilder();

    console.log('Building Gaming Computer:');
    const gamingComputer = director.buildGamingComputer(gamingBuilder);
    gamingComputer.showParts();

    console.log('\nBuilding Office Computer:');
    const officeComputer = director.buildOfficeComputer(officeBuilder);
    officeComputer.showParts();
}

// Testing the Builder Design Pattern
const director = new Director();

console.log('Testing Builder Pattern:\n');
clientCode(director);

File: patterns/Creational/Builder/Go/RealWorldExample/ComputerAssembly.go
package main

import "fmt"

// Builder Design Pattern - Computer Assembly
//
// This example demonstrates the Builder Design Pattern for assembling
// different types of computers with varying configurations (Gaming, Office).

// ComputerBuilder interface declares the product building steps that are common
// for all types of builders.
type ComputerBuilder interface {
    SetCPU(cpu string) ComputerBuilder
    SetRAM(ram string) ComputerBuilder
    SetStorage(storage string) ComputerBuilder
    SetGraphicsCard(gpu string) ComputerBuilder
    GetComputer() Computer
}

// Concrete Builder for assembling a Gaming Computer.
type GamingComputerBuilder struct {
    computer Computer
}

func (b *GamingComputerBuilder) SetCPU(cpu string) ComputerBuilder {
    b.computer.parts["CPU"] = cpu
    return b
}

func (b *GamingComputerBuilder) SetRAM(ram string) ComputerBuilder {
    b.computer.parts["RAM"] = ram
    return b
}

func (b *GamingComputerBuilder) SetStorage(storage string) ComputerBuilder {
    b.computer.parts["Storage"] = storage
    return b
}

func (b *GamingComputerBuilder) SetGraphicsCard(gpu string) ComputerBuilder {
    b.computer.parts["GraphicsCard"] = gpu
    return b
}

func (b *GamingComputerBuilder) GetComputer() Computer {
    return b.computer
}

// Concrete Builder for assembling an Office Computer.
type OfficeComputerBuilder struct {
    computer Computer
}

func (b *OfficeComputerBuilder) SetCPU(cpu string) ComputerBuilder {
    b.computer.parts["CPU"] = cpu
    return b
}

func (b *OfficeComputerBuilder) SetRAM(ram string) ComputerBuilder {
    b.computer.parts["RAM"] = ram
    return b
}

func (b *OfficeComputerBuilder) SetStorage(storage string) ComputerBuilder {
    b.computer.parts["Storage"] = storage
    return b
}

func (b *OfficeComputerBuilder) SetGraphicsCard(gpu string) ComputerBuilder {
    b.computer.parts["GraphicsCard"] = "Integrated GPU" // Office computers don't need high-end GPUs
    return b
}

func (b *OfficeComputerBuilder) GetComputer() Computer {
    return b.computer
}

// Product class representing a Computer.
type Computer struct {
    parts map[string]string
}

func NewComputer() Computer {
    return Computer{parts: make(map[string]string)}
}

func (c *Computer) ShowParts() {
    fmt.Println("Computer Configuration:")
    for key, value := range c.parts {
        fmt.Printf("%s: %s\n", key, value)
    }
}

// Director class guides the building process.
type Director struct{}

func (d *Director) BuildGamingComputer(builder ComputerBuilder) Computer {
    return builder.SetCPU("Intel i9").
        SetRAM("32GB").
        SetStorage("1TB SSD").
        SetGraphicsCard("NVIDIA RTX 3090").
        GetComputer()
}

func (d *Director) BuildOfficeComputer(builder ComputerBuilder) Computer {
    return builder.SetCPU("Intel i5").
        SetRAM("16GB").
        SetStorage("500GB SSD").
        SetGraphicsCard("Integrated GPU").
        GetComputer()
}

// Client code.
func main() {
    director := &Director{}
    gamingBuilder := &GamingComputerBuilder{computer: NewComputer()}
    officeBuilder := &OfficeComputerBuilder{computer: NewComputer()}

    fmt.Println("Building Gaming Computer:")
    gamingComputer := director.BuildGamingComputer(gamingBuilder)
    gamingComputer.ShowParts()

    fmt.Println("\nBuilding Office Computer:")
    officeComputer := director.BuildOfficeComputer(officeBuilder)
    officeComputer.ShowParts()
}
